### PR TITLE
Convert to typescript

### PR DESCRIPTION
**Pull request by usu**
I've started to dive into Typescript. Looks promising to me so far.

@carlobeltrame Can you have a look at it if you feel this goes into the right direction? If so, I'd continue piece-by-piece with the rest of the code.

### CODE REVIEW

**Comment by carlobeltrame** [comment]
Ah wow, nice, thanks! Maybe have a look at https://github.com/ecamp/hal-json-vuex/tree/refactor-to-classes where I started to refactor the storeValueProxy file into classes. I think that should be completed before moving to Typescript, what do you think?

**Comment by usu** [comment]
> I think that should be completed before moving to Typescript, what do you think?
Ah nice, didn't see that branch before. Yes, makes sense to complete & merge that one before. Otherwise we'll have tons of conflicts.

File: src/StoreValue.ts
Comment (carlobeltrame) [inline src/StoreValue.ts:39]: I agree, can you remove the line and check if the tests still run?
This was added when `allItems` was [introduced](https://github.com/ecamp/ecamp3/commit/7ec5a67ff7e9226f4bf044f00e89e345e1a422db), might have been a debugging leftover

File: src/QueryablePromise.ts
Comment (carlobeltrame) [inline src/QueryablePromise.ts:41]: Maybe not part of this PR, but shouldn't that also return a `QueryablePromise`? The properties of this returned `Promise` are writable, while the `QueryablePromise` properties only have getters.
And we have ternary operators that either return a `new QueryablePromise` or `QueryablePromise.resolve(...)`, so TypeScript could be more certain about the outcome there.

File: src/QueryablePromise.ts
Comment (carlobeltrame) [inline src/QueryablePromise.ts:6]: Should this extend `Promise<T>`? Or is that not possible?

File: src/QueryablePromise.ts
Comment (carlobeltrame) [reply src/QueryablePromise.ts:41]: Nevermind, as soon as we use `.then(...)` it reverts back to a normal `Promise`. Can we override `then` and `catch` in some way so the resulting Promise is also queryable, or is that a bad idea?

File: src/interfaces/ApiActions.ts
Comment (carlobeltrame) [inline src/interfaces/ApiActions.ts:7]: Could we require `data` to be of `object` type?

File: src/interfaces/Config.ts
Comment (carlobeltrame) [inline src/interfaces/Config.ts:15]: I didn't find any uses of standalone `ExternalConfig`, is this still WIP?

File: webpack.config.js
Comment (carlobeltrame) [inline webpack.config.js:34]: Is this so the source maps can point to the *.ts files? If so, do source maps work for you so far? I think we need to include the source code in the NPM package before the source maps can work (see .npmignore)

File: src/interfaces/StoreData.ts
Comment (carlobeltrame) [inline src/interfaces/StoreData.ts:6]: There are more known possible entries here, such as `loading?: boolean`, `deleting?: boolean` `reloading?: boolean` and `reload?: { uri: string, property: string }`

**Review by carlobeltrame — CHANGES_REQUESTED** [review]
Looking good :) a few comments from my side for now. Let me know if you want me to continue this.

File: src/interfaces/Config.ts
Comment (usu) [reply src/interfaces/Config.ts:15]: Yep, WIP. `ExternalConfig` will specify the config which users can provide when installing `hal-json-vuex`

File: src/interfaces/StoreData.ts
Comment (usu) [reply src/interfaces/StoreData.ts:6]: Yep also WIP. Will add them when migrating `index.js`. The code converted so far only uses `self` and `load` and I didn't want to add properties in advance (kind of testing on the fly that TypeScript finds the issue and complains as expected).

File: webpack.config.js
Comment (usu) [reply webpack.config.js:34]: Honestly, didn't use or test this so far. Was just a copy&paste from a Typescript/Webpack guide.
I'll keep this open here until we have a better understanding, what needs to be included, such that debugging is possible for users of the package.

File: src/QueryablePromise.ts
Comment (usu) [reply src/QueryablePromise.ts:41]: You're right. `storeDate._meta.load` should be of type `QueryablePromise` and all methods modifying `load` should ensure it's of this type. Extending `Promise` and overriding `then`/`catch` could be an option. Need to give it a few thoughts.

Comment (carlobeltrame) [reply src/QueryablePromise.ts:41]: And `QueryablePromise.resolve` should also return `QueryablePromise` in that case

File: src/interfaces/ApiActions.ts
Comment (usu) [reply src/interfaces/ApiActions.ts:7]: I had this before, then eslint complained:

```
Don't use `object` as a type. The `object` type is currently hard to use ([see this issue](https://github.com/microsoft/TypeScript/issues/21732)).
Consider using `Record<string, unknown>` instead, as it allows you to more easily inspect and use the keys.eslint@typescript-eslint/ban-types)
```

Which made me thinking if we really want to limit axios payload to objects? Axios has `any` as payload type (which actually is not encouraged by Typescript). At the moment, we may only post/patch JSON objects, but maybe someone wants to send binary data to the server, too. AFAIK HAL JSON only specifies the receiving part, but says nothing about the format when sending payload.

**Comment by usu** [comment]
@carlobeltrame I believe this is in a state where you could have another look at it. Would say 95% complete as a first shot. Appreciate a careful review.

Some open actions (potentially as separate issues, to be discussed):
- [ ] Discuss and resolve the various inline comments marked as "TODO"
- [ ] Avoid usage of "any" and inline rule exclusions
- [ ] Real world test in ecamp3 (only tested independetly so far)
- [ ] Consider splitting up index (Install/Export, API actions, helper functions)
- [ ] Consider implementation of CanHaveItems as mixin/composable instead of as a base class
- [ ] Naming and file organization: Names of classes/interfaces/etc. have now grown historically. I think it makes sense to take a step back and review what the best and most consistent naming would be.
- [ ] `LoadingStoreCollection` is a class with a static function only which is a Javascript antipattern. Consider implementing as simple function or integrate back to `LoadingStoreValue`.
- [ ] Unit tests for separate components?

File: src/QueryablePromise.ts
Comment (carlobeltrame) [inline src/QueryablePromise.ts:81]: How about you just do this:
```suggestion
  return wrapPromise(new Promise(resolve => resolve(value)))
```

File: src/StoreValue.ts
Comment (carlobeltrame) [inline src/StoreValue.ts:16]: We should probably move the `@param` annotations to the constructor functions

**Review by carlobeltrame — CHANGES_REQUESTED** [review]
Nice work, thank you! I have tried to answer all your todos. There is still some work to do before we can merge this I think, but if you feel that some things should be done in later iterations let me know.

**Comment by usu** [comment]
> There is still some work to do before we can merge this I think, but if you feel that some things should be done in later iterations let me know.

Yes agree. Let me see how far I get with the open points. Some of the questions I raised in the earlier comment might be too big to also squeeze into this pull request.

By the way: I also just uploaded a [rebased version](https://github.com/usu/hal-json-vuex/commits/feature/typescript-rebased) of the code status as of today (separating renaming of files and actual code change). Might be a bit easier to verify the code changes by looking at the last commit only:
https://github.com/usu/hal-json-vuex/commit/17d57ff7977a3a2916d3d7332d5095e7568a8ca4

File: webpack.config.js
Comment (usu) [reply webpack.config.js:34]: For the first time, I tried to run ecamp3 with this branch of `hal-json-vuex`.

**The bad thing:** Something is clearly not working properly, error in console & no data loaded at all

**The good thing:** The error message in the browser directly lead me to to the correct source code line in `index.ts`. So the source map seems to work. Looks like `bundle.js.map` includes the complete original Typescript code.

Comment (carlobeltrame) [reply webpack.config.js:34]: Nice. How did you run it with this branch? Did you `npm run build` and then copy only the files from `dist` over to ecamp3?

Comment (usu) [reply webpack.config.js:34]: Sort of, but more automated. I tried first with `npm link`, but docker obviosuly doesn't like it because the symlink will be outside of docker's volume.

To the rescue:
https://github.com/whitecolor/yalc

In `hal-json-vuex`:
```
npm run build
sudo yalc publish
```

in `ecamp3/frontend`:
```
yalc add hal-json-vuex
```

and later
```
yalc update
```

I'm quite sure I managed to run it with `yalc link hal-json-vuex` earlier. But somehow it doesn't work anymore. Theoritically it should work because the symlink stays within `frontend` and therefore within the docker volume.

File: src/LoadingStoreValue.ts
Comment (usu) [reply src/LoadingStoreValue.ts:62]: I added `toString()` a few lines below as a function of `LoadingStoreValue`... which obviosuly is total nonsense. Will add it back 😄

File: src/CanHaveItems.ts
Comment (usu) [reply src/CanHaveItems.ts:103]: Can you make an example of what that would like? Like an API response that uses `items` as a normal property containing an array? Like this:

```yaml
{
    "id": 1,
    "items": [
        1,
        2,
        3
    ],
    "_links": {
        "self": {
            "href": "/camps/1"
        }
    }
}
```

File: src/EmbeddedCollection.ts
Comment (usu) [reply src/EmbeddedCollection.ts:51]: What's the expectation on what $loadItems returns? A Promise which resolves to items when all items are loaded from the API?

File: src/CanHaveItems.ts
Comment (carlobeltrame) [reply src/CanHaveItems.ts:103]: For example. Or even completely mixed, I'm not sure whether this is allowed in HAL:
```json
{
  "id": 1,
  "_embedded": {
    "items": [
      1,
      { "href": "/entries/2" },
      3
    ]
  },
  "_links": {
    "self": {
      "href": "/camps/1"
    }
  }
}
```

File: src/EmbeddedCollection.ts
Comment (carlobeltrame) [reply src/EmbeddedCollection.ts:51]: Yes, exactly.

File: src/LoadingStoreCollection.ts
Comment (usu) [reply src/LoadingStoreCollection.ts:12]: In this case, we either need to wrap it. Or extend the Array class. I tried latter, and somehow failed. Let my try former :dizzy_face:

File: src/CanHaveItems.ts
Comment (usu) [reply src/CanHaveItems.ts:50]: Mixin pattern (or composable) is basically an alternative to OOP inheritance. See here:
https://www.typescriptlang.org/docs/handbook/mixins.html
https://mariusschulz.com/blog/mixin-classes-in-typescript
https://www.digitalocean.com/community/tutorials/typescript-mixins

In your code  example above you actually just described the mixin mattern (not inheriting anymore but applying properties, methods, etc. on top on an existing object).

What I'd do differently: I would escalate the triage `isCollection` or not to the StoreValueCreator. Thats's the only place where StoreValues are created and at that point, we could already decide whether we need a Collection or not.

Something like this:

- `Collection` class = extends `Resource` base class + `CanHaveItems` mixin on top
- `Entity` class = extends `Resource` base class with nothing on top (at least nothing I could see right now)

StoreValueCreator (or better ResourceCreator?) does the triage:
- Still loading --> you get a LoadingResource
- Loaded and has 'items' --> you get a Collection
- Loaded and no 'items' --> you get an Entity

Comment (carlobeltrame) [reply src/CanHaveItems.ts:50]: Sounds good! And then we can really get rid of LoadingCollection.

File: src/LoadingStoreValue.ts
Comment (usu) [reply src/LoadingStoreValue.ts:39]: The user of the package still knows because the original load promise is returned back (e.g. via patch(), post(), etc.).

I had a situation where my tests failed randomly and it took me quite some time to track it down to this. Unfortunately I cannot reproduce now anymore, so removing `.catch(() => this)` right now would still pass all tests.

What happens is that every `.then()` on a promise creates a new promise. If the original promise throws an error, every promise created by `.then()` potentially throws one, too.

See example here:
https://jsfiddle.net/r457jqm2/

If you run this example you should see one console output "Crap", but also 3 unhandled promise errors. This might be a nuisance only in browser environment. However node doesn't like this. In best case, we get a "Unhandled promise rejection" warning. In worst case, tests fail that are completely unrelated to this code because they get failing promises from previous test runs.

File: src/QueryablePromise.ts
Comment (usu) [reply src/QueryablePromise.ts:81]: QueryablePromise is not needed anymore after this commit:
https://github.com/ecamp/hal-json-vuex/pull/9/commits/48bc1b95c6d07b42ad9367305f972dd4a268085a

File: src/QueryablePromise.ts
Comment (usu) [reply src/QueryablePromise.ts:55]: QueryablePromise is not needed anymore after this commit:
https://github.com/ecamp/hal-json-vuex/pull/9/commits/48bc1b95c6d07b42ad9367305f972dd4a268085a

File: src/LoadingStoreValue.ts
Comment (usu) [reply src/LoadingStoreValue.ts]: Removed with this commit:
https://github.com/ecamp/hal-json-vuex/pull/9/commits/8365df6908aefd97290c55b413ade227193aee01

File: src/QueryablePromise.ts
Comment (carlobeltrame) [reply src/QueryablePromise.ts:55]: Aha! Nice work! I think you are right, the reloading check prompted me to introduce the done flag on the promise in the first place.

File: src/HasItems.ts
Comment (usu) [inline src/HasItems.ts:77]: Linter doesn't like this:
`'HasItems' is defined but never use`

**Comment by carlobeltrame** [comment]
I have also continued this a little, and I have come to the conclusion that for HasItems, normal inheritance should be enough, we don't need the more generic variant of mixins you described (a mixin is valuable if you want to apply it to multiple different base classes, but we only ever apply it to StoreValue).
But I am not quite finished with resolving all type-related refactorings I was doing. Maybe we can have a look at it some time.

**Comment by carlobeltrame** [comment]
Mein aktuellster Stand ist jetzt auf diesem Branch: https://github.com/ecamp/hal-json-vuex/tree/feature/typescript

File: src/LoadingStoreCollection.ts
Comment (carlobeltrame) [reply src/LoadingStoreCollection.ts:12]: One conclusion from my experiments: Extending the Array class is not possible, because the LoadingStoreCollection has different type signatures for either `find` or `map`:
* If `LoadingStoreCollection` inherits from `Array<Resource>` or some other `Array<Something>`, or we make a generic `LoadingStoreCollection<T>`, the `find` method can no longer return a `LoadingStoreValue`, but must return `Resource` or `Something` or `T`.
* If `LoadingStoreCollection` inherits from `Array<LoadingStoreValue>`, the `map` method cannot be made to work the way we want, because it has to potentially transform the `LoadingStoreValue`s into objects of other classes, depending on the callback passed to `map` (the callback might map every entry to an integer, for example).

Comment (carlobeltrame) [reply src/LoadingStoreCollection.ts:12]: TODO: Try something like `LoadingStoreValue<T> extends T`, this way the Array contract could maybe hold

File: src/HasItems.ts
Comment (carlobeltrame) [reply src/HasItems.ts:40]: We'll leave this as it is for now, maybe deprecate `$loadItems` in the future once we get the implementation right.

File: src/index.ts
Comment (usu) [reply src/index.ts]: Fixed with 40948f9

File: src/index.ts
Comment (usu) [reply src/index.ts]: Fixed with 40948f9

**Comment by usu** [comment]
@carlobeltrame 
The 2 open points discussed last week are now implemented
- `$loadItems` resolves to the collection object instead of the array of items
- `get` cannot be used anymore with embedded collections (limited to reloading)

We'll probably still see more changes on the embedded collections. I don't think it's coded super cleanly.

But would appreciate if we can merge the current status and then divide future work into smaller chunks.

Alternatively, we could also open a v2-alpha branch (+v2 alpha releases) and start PR/merging into the alpha branch until we feel  the interface is robust and stable enough.

File: src/EmbeddedCollection.ts
Comment (carlobeltrame) [inline src/EmbeddedCollection.ts:1]: Remove?

File: src/EmbeddedCollection.ts
Comment (carlobeltrame) [inline src/EmbeddedCollection.ts:31]: Outdated jsDoc comment

File: src/HasItems.ts
Comment (carlobeltrame) [inline src/HasItems.ts:40]: Indentation

File: src/HasItems.ts
Comment (carlobeltrame) [inline src/HasItems.ts:79]: This `as` is here to convince TypeScript that the server has previously sent a collection, so it will send a collection again, right? Consider adding a comment that explains this

**Review by carlobeltrame — APPROVED** [review]
LGTM, thanks a lot!

The alpha strategy sounds good. In case we ever decide to backport anything to an old major version, I'd rather have a sound pre-typescript version than an early typescript version which may contain flaws that we do not know as of now.

**Comment by usu** [comment]
@carlobeltrame I've changed the target branch to `next` instead of `master`. So we merge and then tag/publish as `2.0.0-alpha.0`? Or `1.3.0-alpha.0`?

**Comment by carlobeltrame** [comment]
Yes, +1 for 2.0.0-alpha.0, since technically we already made some small breaking changes (to undocumented features), and this allows us to clean up some more if we deem it necessary.

**Comment by carlobeltrame** [comment]
v2.0.0-alpha.0 is published: https://www.npmjs.com/package/hal-json-vuex?activeTab=versions

**Pull request merged**
